From: Christopher Obbard <chris.obbard@collabora.com>
Date: Fri, 10 Jul 2020 11:12:16 +0100
Subject: usb: xhci: apply suspend quirks to AMD XHCI controller 1022:15e5

This controller timeouts during suspend (S3) from a warm-boot with
[   46.011685] xhci_hcd 0000:03:00.3: WARN: xHC save state timeout
[   46.011698] PM: suspend_common(): xhci_pci_suspend+0x0/0xd0 [xhci_pci] returns -110
[   46.011713] PM: pci_pm_suspend(): hcd_pci_suspend+0x0/0x30 [usbcore] returns -110
[   46.011717] PM: dpm_run_callback(): pci_pm_suspend+0x0/0x130 returns -110
[   46.011719] PM: Device 0000:03:00.3 failed to suspend async: error -110
[   46.180504] PM: Some devices failed to suspend, or early wake event detected
[   46.180681] hub 1-0:1.0: hub_ext_port_status failed (err = -108)
[   46.180686] usb usb1-port1: cannot disable (err = -108)
[   46.180705] hub 1-0:1.0: hub_ext_port_status failed (err = -108)
[   46.180708] usb usb1-port2: cannot disable (err = -108)
[   46.180717] hub 1-0:1.0: hub_ext_port_status failed (err = -108)
[   46.180720] usb usb1-port4: cannot disable (err = -108)
[   46.180729] hub 1-0:1.0: hub_ext_port_status failed (err = -108)
[   46.180734] usb usb1-port5: cannot disable (err = -108)
thus preventing the system from entering S3.

It remains in an undefined state and connected devices stop working
until a reboot. Since this causes an early wake, other devices connected
to the system (e.g. GPU) do not wakeup properly and these other devices
are also in an undefined state.

Apply XHCI_SUSPEND_DELAY and XHCI_SNPS_BROKEN_SUSPEND quirks to allow
it to suspend properly.

Signed-off-by: Christopher Obbard <chris.obbard@collabora.com>
---
 drivers/usb/host/xhci-pci.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/host/xhci-pci.c b/drivers/usb/host/xhci-pci.c
index 4917c5b..a6b940e 100644
--- a/drivers/usb/host/xhci-pci.c
+++ b/drivers/usb/host/xhci-pci.c
@@ -137,11 +137,14 @@ static void xhci_pci_quirks(struct device *dev, struct xhci_hcd *xhci)
 	if (pdev->vendor == PCI_VENDOR_ID_AMD &&
 		(pdev->device == 0x15e0 ||
 		 pdev->device == 0x15e1 ||
+		 pdev->device == 0x15e5 ||
 		 pdev->device == 0x43bb))
 		xhci->quirks |= XHCI_SUSPEND_DELAY;
 
 	if (pdev->vendor == PCI_VENDOR_ID_AMD &&
-	    (pdev->device == 0x15e0 || pdev->device == 0x15e1))
+		(pdev->device == 0x15e0 ||
+		 pdev->device == 0x15e1 ||
+		 pdev->device == 0x15e5))
 		xhci->quirks |= XHCI_SNPS_BROKEN_SUSPEND;
 
 	if (pdev->vendor == PCI_VENDOR_ID_AMD)
